src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clear(stateCopy);
+        break;
+
+      default:
+        return 'Error';
+    }
+
+    result.push({ ...stateCopy });
+  }
+
+  return result;
+}
+
+function addProperties(state, extraData) {
+  Object.assign(state, extraData);
+}
```

Comment:
[CODE QUALITY]: The `addProperties` function directly modifies the passed `state` object. This is fine within the context of this task because you're working with a copy of the state, but it's important to note that as a standalone function, this would be mutating the input object. Consider returning a new object instead if this function is to be used in a different context.